DAOS-5195 object: Choose epochs on servers

liw opened the pull request:
Transaction epochs are currently chosen on clients. To avoid depending
client HLCs, this patch chooses transaction epochs on servers. Before a
transaction's epoch has been chosen, one task, among possibly multiple
concurrent tasks associated with this transaction, chooses an epoch when
executing on a server; the other tasks wait for this "epoch task" to
complete. Below is a list of the changes:

  - dtx_epoch
      - Rename dc_obj_epoch to dtx_epoch, so that this type can be used
        on the server side.
      - Add oe_first, so that dtx_id.dti_hlc, which is based on client
        HLCs, will no longer be used to find the first epoch of a TX.
  - The RPC protocol
      - Add epoch_first fields to read requests, for passing
        dtx_epoch.oe_first to servers.
      - Add epoch fields to read replies, for passing chosen epochs to
        clients.
  - dc_tx
      - No longer chooses epochs using client HLCs.
      - Add tx_epoch_task, for tracking tasks that are choosing epochs
        for TXs.
      - Add dc_tx_get_epoch, for tasks to choose or wait for epochs.
  - Remove the local TX from dsc_obj_list_obj, as we already specify an
    epoch range.

Signed-off-by: Li Wei <wei.g.li@intel.com>

liw commented:
Only requesting PRELIMINARY reviews, since the PR has never been tested by CI yet. I have run "daos_test -mpceXVizADKCoROFNv" manually with 3 servers. All tests can pass, except for EP1, which abuses TX API.

daosbuild1 commented on src/object/srv_cli.c:
```diff
@@ -179,27 +179,14 @@ dsc_obj_list_obj(daos_handle_t oh, daos_epoch_range_t *epr, daos_key_t *dkey,
-		return rc;
-
-	rc = dc_obj_list_obj_task_create(oh, th, epr, dkey, akey, size, nr,
+	rc = dc_obj_list_obj_task_create(oh, DAOS_TX_NONE, epr, dkey, akey, size, nr,
```
(style)  line over 80 characters

daosbuild1 commented on src/object/obj_tx.c:
```diff
@@ -445,6 +466,106 @@ dc_tx_hdl2epoch_and_pmv(daos_handle_t th, struct dc_obj_epoch *epoch,
+	if (tx->tx_epoch_task == task) {
+		tse_task_decref(tx->tx_epoch_task);
+		tx->tx_epoch_task = NULL;
+		D_DEBUG(DB_IO, DF_X64"/%p: epoch task complete\n", th->cookie, task);
```
(style)  line over 80 characters

daosbuild1 commented on src/object/obj_tx.c:
```diff
@@ -445,6 +466,106 @@ dc_tx_hdl2epoch_and_pmv(daos_handle_t th, struct dc_obj_epoch *epoch,
+
+	tx = dc_tx_hdl2ptr(*th);
+	if (tx == NULL) {
+		D_ERROR("cannot find transaction handle "DF_X64"\n", th->cookie);
```
(style)  line over 80 characters

daosbuild1 commented on src/object/cli_obj.c:
```diff
@@ -2149,10 +2157,29 @@ shard_io(tse_task_t *task, struct shard_auxi_args *shard_auxi)
 
+	/*
+	 * If this task belongs to a TX, and if the epoch we got earlier
+	 * doesn't contain a "chosen" TX epoch, then we may need to reinit the
```
(style)  'doesn'' may be misspelled - perhaps 'doesn't'?

daosbuild1 commented on src/object/cli_obj.c:
```diff
@@ -2031,8 +2040,7 @@ shard_task_sched(tse_task_t *task, void *arg)
 		    obj_retry_error(task->dt_result) ||
-		    sched_arg->tsa_epoch.oe_value !=
-		    shard_auxi->epoch.oe_value ||
+		    dtx_epoch_equal(&sched_arg->tsa_epoch, &shard_auxi->epoch) ||
```
(style)  line over 80 characters

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3108/3/execution/node/58/log

liw commented:
Changes since last revision:

- Rebase to the latest master.
- Fix EP1 (i.e., "daos_test -e") to stop using committed TX handles for fetching.
- Support daos_tx_hdl2epoch for non_cpd UPDATEs. (PUNCHs will only work after cpd is available.)

liw commented:
This PR is ready for full reviews, please.

daosbuild1 commented on src/tests/suite/daos_epoch.c:
```diff
@@ -97,15 +95,23 @@ io_for_aggregation(test_arg_t *arg, daos_handle_t coh, daos_handle_t ths[],
-		}
-		lookup_single(dkey, akey, 1100, rec, REC_MAX_LEN, ths[i], &req);
+		if (verify_data == NULL)
+			snprintf(rec_verify, REC_MAX_LEN, VAL_FMT, (unsigned long)i);
```
(style)  line over 80 characters

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3108/4/execution/node/59/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3108/4/execution/node/809/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-3108/4/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3108/4/execution/node/746/log

Nasf-Fan commented on src/object/cli_obj.c:
```diff
@@ -2256,7 +2288,7 @@ obj_req_fanout(struct dc_object *obj, struct obj_auxi_args *obj_auxi,
 			}
 			goto task_sched;
-		} else {
+		} else if (!require_shard_task) {
```
How to handle a retry RPC with "require_shard_task" as true?

liw replied:
It's not clear indeed. I'm not sure about what the best approach is. Shall we discuss this?

liw replied:
We had a discussion on this. I'll think further, especially considering the new development related to HLCT.

Nasf-Fan commented on src/object/cli_obj.c:
```diff
@@ -2149,10 +2158,28 @@ shard_io(tse_task_t *task, struct shard_auxi_args *shard_auxi)
+	 * shard_io_task.
+	 */
+	th = shard_auxi->obj_auxi->th;
+	if (daos_handle_is_valid(th) && !dtx_epoch_chosen(&shard_auxi->epoch)) {
```
Only for fetch ? For update and punch cases, we just cache on the client without consider the epoch before TX commit.

liw replied:
I needed to remove the "is-a-read-operation" condition for non_cpd cases in EP1. I think the removal is harmless because for cpd cases update/punch operations should not reach this code path, while CPD (if it goes through here) should get choosing and later set epoch for daos_tx_hdl2epoch calls after commit.

Nasf-Fan commented on src/object/cli_shard.c:
```diff
@@ -300,6 +302,22 @@ dc_rw_cb(tse_task_t *task, void *arg)
+	/*
+	 * orwo->orw_epoch may be set even when the status is nonzero (e.g.,
+	 * -DER_TX_RESTART and -DER_INPROGRESS). If this task is a TX "epoch
+	 * task", we shall call dc_tx_set_epoch.
```
Is is possible that the orw_epoch is not initialized if the returned RPC status is non-zero?

liw replied:
Uninitialized orw_epoch == 0.

Nasf-Fan commented on src/object/obj_tx.c:
```diff
@@ -864,19 +987,14 @@ dc_tx_restart(tse_task_t *task)
-		tx->tx_epoch.oe_value = crt_hlc_get();
+		tx->tx_status = TX_OPEN;
+		if (dtx_epoch_chosen(&tx->tx_epoch))
+			tx->tx_epoch.oe_flags |= DTX_EPOCH_HINT;
```
dc_tx_restart() may be triggered by the TX sponsor if some DAOS API (such as fetch) return "-DER_TX_RESTART", but at that point, the fetch RPC reply may already contained the valid epoch and set to the tx_epoch via dc_tx_set_epoch(). So here, if we mark the tx_epoch.oe_flags as "DTX_EPOCH_HINT", does it means we need another fetch to refetch another epoch from the server after the restart?

liw replied:
Yes, the restarted TX begins without a chosen epoch. If the user calls FETCH first, this FETCH chooses an epoch; if the user calls UPDATE first, the CPD chooses an epoch, etc. The HINT flag ensures that any new epoch chosen will be higher than the previous chosen epoch.

liw commented on src/object/obj_tx.c:
```diff
@@ -445,6 +466,108 @@ dc_tx_hdl2epoch_and_pmv(daos_handle_t th, struct dc_obj_epoch *epoch,
+					       sizeof(th));
+		if (rc != 0) {
+			D_ERROR("cannot register completion callback: "DF_RC
+				"\n", DP_RC(rc));
```
Indeed. Fixed.

liw commented on src/object/obj_tx.c:
```diff
@@ -445,6 +466,108 @@ dc_tx_hdl2epoch_and_pmv(daos_handle_t th, struct dc_obj_epoch *epoch,
+				"\n", DP_RC(rc));
+			goto out;
+		}
+		*epoch = tx->tx_epoch;
```
In this case, the task still needs to send a 0, MAX, or hint epoch to the server. dc_tx_get_epoch initializes *epoch for this purpose.

liw commented on src/object/srv_obj.c:
```diff
@@ -1617,6 +1634,55 @@ obj_tgt_update(struct dtx_leader_handle *dlh, void *arg, int idx,
+		 * epoch higher than the epochs of the operations completed on
+		 * this server.
+		 */
+		crt_hlc_get_msg(*epoch);
```
Suppose this is server X, whose HLC "runs behind others". The epoch may be chosen by server Y, whose HLC "runs ahead of" server X. If we don't sync server X's HLC here, next time we get server X's HLC reading, the resulting HLC timestamp may be lower than the epoch used by this I/O.

That said, I'm discussing some changes with Dmitry on how to use the hlc fields in crt common hdr. Depending on how that discussion goes, I may remove this sync if we choose to do it in crt level. At this moment, the sync is necessary.

liw commented:
Changes since last revision:

- Rebase to latest master.
- Remove epoch hints, relying on HLC (for now) and HLCT (for future).
- Set TX to TX_FAILED upon -DER_TX_RESTART.
- Handle obj_req_fanout case "retry && require_shard_task".
- Fix task ref leak in dc_tx_get_epoch.

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-3108/5/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3108/5/execution/node/842/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3108/5/execution/node/768/log

liw commented:
Functional_Hardware_Small hdf5: DAOS-5321
Functional_Hardware_Medium daos_test -d: DAOS-5344
Functional_Hardware_Large: no test failure, but the result archiving met "no space" error.

liw commented:
Changes since last revision:

- Rebased to latest master.
- Removed TX status changes, deferring to separate PR.

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-3108/6/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3108/6/execution/node/751/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3108/6/execution/node/779/log

liw commented:
Functional_Hardware_Small hdf5: DAOS-5321
Functional_Hardware_Medium daos_test -d: DAOS-5344
Functional_Hardware_Large: no test failure, but the result archiving met "no space" error. (Seems to happen with other master runs too.)

mchaarawi commented on src/object/cli_obj.c:
```diff
@@ -2198,12 +2225,17 @@ obj_req_fanout(struct dc_object *obj, struct obj_auxi_args *obj_auxi,
 	struct shard_auxi_args	*shard_auxi;
 	struct daos_shard_tgt	*tgt;
 	uint32_t		 i, tgts_nr;
+	bool			 require_shard_task = false;
```
so when we have a fanout of enumerate or key_query task, do we need to have 1 task go first and update the other tasks?

liw replied:
Currently, when such an operation is invoked with DAOS_TX_NONE, their fanout RPCs still go concurrently. (These fanout RPCs may effectively see an inconsistent set of results. We would need to open an implicit TX to guarantee consistency.) When an operation is invoked with a TX handle, yes, one of the fanout RPCs goes first, and the others need to wait for this first RPC to choose an epoch. See shard_io_task and shard_query_key_task.

mchaarawi replied:
i see. could you please point me to where we add the dependency on this first task that goes ahead? maybe it was done in a previous patch and i missed it.
For the TX_NONE case, I think doing an implicit TX on the client side and having to wait for one shard to go ahead first might be bad for performance in that mode. but i know this is unrelated to this PR.

liw replied:
For dependency, it is in this PR. See shard_io_task and shard_query_key_task, which call dc_tx_get_epoch, who handles the dependency (tracked by dc_tx.tx_epoch_task).

For TX_NONE and fanout, it depends on the consistency we support. If we do need to support consistency, I'm afraid there's few options.

mchaarawi replied:
ah i see it now. sorry about that!

liw replied:
No problem at all. The PR isn't small.

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3108/7/execution/node/59/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-3108/7/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3108/7/execution/node/764/log

liw commented:
Functional_Hardware_Small hdf5: DAOS-5321
Functional_Hardware_Large osa_online_reintegration: DAOS-5383

liw commented:
I had to merge master again to resolve some recent conflicts; no changes otherwise.

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3108/8/execution/node/59/log

Nasf-Fan commented on src/object/cli_obj.c:
```diff
@@ -2150,8 +2159,8 @@ shard_task_sched(tse_task_t *task, void *arg)
-		    sched_arg->tsa_epoch.oe_value !=
-		    shard_auxi->epoch.oe_value ||
+		    dtx_epoch_equal(&sched_arg->tsa_epoch,
+				    &shard_auxi->epoch) ||
```
"! dtx_epoch_equal(&sched_arg->tsa_epoch, &shard_auxi->epoch)"?

Nasf-Fan commented on src/object/cli_obj.c:
```diff
@@ -2268,10 +2277,28 @@ shard_io(tse_task_t *task, struct shard_auxi_args *shard_auxi)
+		rc = dc_tx_get_epoch(task, th, &shard_auxi->epoch);
+		if (rc == DC_TX_GE_REINIT)
+			return tse_task_reinit(task);
+		else if (rc < 0)
```
"else" is unnecessary since above "if" branch have returned.

liw replied:
I meant to write this as one "if ... else if ... else if ..." block for checking rc. I think breaking it into two "if" blocks would not make things a bit worse. And there's no compelling reason to break it, is there?

Nasf-Fan commented on src/object/srv_obj.c:
```diff
@@ -1606,6 +1621,39 @@ obj_tgt_update(struct dtx_leader_handle *dlh, void *arg, int idx,
 
+/* Nonnegative return codes of process_epoch */
+enum process_epoch_rc {
+	PE_OK_REMOTE,	/* OK and epoch chosen remotely */
```
Who will use this return value? There is no use case in this patch.

liw replied:
We check if the rc is PE_OK_LOCAL. If it's not LOCAL, then it's REMOTE. Having a name is better than checking if rc is 1, for instance.

Nasf-Fan replied:
Since then, only "PE_OK_LOCAL" seems enough.

liw commented:
Addressed review comments and merged with the latest master.

liuxuezhao commented on src/object/cli_obj.c:
```diff
@@ -2375,6 +2407,21 @@ obj_req_fanout(struct dc_object *obj, struct obj_auxi_args *obj_auxi,
+			 * aborting and waiting for associated operations when
+			 * restarting a TX.
+			 */
+			return -DER_OP_CANCELED;
```
just curious did you meet this case in test? how could it happen, an OP in a TX, retried and then find no epoch set for it...

liw replied:
No, it would be quite difficult to trigger. Fan Yong and I came up with this case by reasoning. There could be a dc_tx_restart immediately before the retry.

liw commented on src/object/cli_obj.c:
```diff
@@ -1883,6 +1883,15 @@ obj_req_valid(tse_task_t *task, void *args, int opc, struct dc_obj_epoch *epoch,
+			    l_args->eprs != NULL) {
+				D_ERROR("mutually exclusive th and eprs "
+					"specified for listing objects\n");
+				rc = -DER_INVAL;
```
In the DAOS_OBJ_RPC_ENUMERATE case, eprs is supposed to pass in a caller-specified epoch range. For operations belonging to a th, the th decides their epochs. So mixing the two usages in one DAOS_OBJ_RPC_ENUMERATE call makes little sense.

Right, dsc_obj_list_obj does this. This PR fixes dsc_obj_list_obj to not use a th.

liuxuezhao replied:
i see, thanks. Sorry did not see the change in dsc_obj_list_obj yesterday.

liw replied:
No problem at all. The PR isn't a small one.

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-3108/9/testReport/(root)/

liw commented:
Conflicts again. Merged with the latest master; no change.

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3108/10/execution/node/803/log

liw commented:
Functional_Hardware_Medium rebuild interruptions: DAOS-5362 & DAOS-5371 (i.e., also happens without this PR)
